pkg/io/v2/providers/providers.go

```diff
@@ -0,0 +1,215 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package providers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"gocloud.dev/blob"
+	_ "gocloud.dev/blob/fileblob"
+	"gocloud.dev/blob/gcsblob"
+	"gocloud.dev/blob/s3blob"
+	"gocloud.dev/gcp"
+	"golang.org/x/oauth2/google"
+)
+
+const (
+	// defaults to file because local paths were
+	// previously given without file:// prefix
+	defaultStorageProviderName = "file"
+
+	storageSeparator = "://"
+
+	providerFile = "file"
+	providerGS   = "gs"
+	providerS3   = "s3"
+)
+
+var storageProviders = []string{providerFile, providerGS, providerS3}
+
+// GetBucket opens and returns a gocloud blob.Bucket based on credentials and a path.
+// The path is used to discover which storageProvider should be used.
+//
+// If the storageProvider file is detected, we don't need any credentials and just open a file bucket
+// If no credentials are given, we just fall back to blob.OpenBucket which tries to auto discover credentials
+// e.g. via environment variables. For more details, see: https://gocloud.dev/howto/blob/
+//
+// If we specify credentials and an gs:// or s3:// path is used, credentials must be given in one of the
+// following formats:
+// * Google Cloud Storage (gs://):
+//    {
+//      "type": "service_account",
+//      "project_id": "<project_id>",
+//      "private_key_id": "<private_key_id>",
+//      "private_key": "<private_key>",
+//      "client_email": "<client_email>",
+//      "client_id": "<client_id>",
+//      "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+//      "token_uri": "https://oauth2.googleapis.com/token",
+//      "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
+//      "client_x509_cert_url": "<client_x509_cert_url>"
+//    }
+// * AWS S3 (s3://):
+//    {
+//      "region": "us-east-1",
+//      "s3_force_path_style": true,
+//      "access_key": "access_key",
+//      "secret_key": "secret_key"
+//    }
+// * S3-compatible service, e.g. self-hosted Minio (s3://):
+//    {
+//      "region": "minio",
+//      "endpoint": "https://minio-hl-svc.minio-operator-ns:9000",
+//      "s3_force_path_style": true,
+//      "access_key": "access_key",
+//      "secret_key": "secret_key"
+//    }
+func GetBucket(ctx context.Context, credentials []byte, path string) (*blob.Bucket, error) {
+	storageProvider, bucket, _, err := ParseStoragePath(path)
+	if err != nil {
+		return nil, err
+	}
+	if storageProvider == providerFile || len(credentials) == 0 {
```

Comment:
I think we should remove the first part of the condition, passing in credentials and using the file provider is invalid. Also, nitpicky comment, I find it easier to read if the `len(credential) == 0` case is at the bottom of the func, because its the default we fall through to.

Reply:
Done